addons/edi_sale/models/edi_sale_request_document.py

```diff
@@ -95,13 +95,32 @@ def prepare(self, doc):
     @api.model
     def execute(self, doc):
         """Execute document"""
+        Partner = self.env['res.partner']
+
         super().execute(doc)
 
+        # Delete sale orders which no longer have any order lines.
+        SaleRequestRecord = self.sale_request_record_model(doc)
+        reqs = SaleRequestRecord.search([('doc_id', '=', doc.id)])
+        sales = reqs.mapped('sale_id')
+        to_remove = sales.filtered(lambda s: len(s.mapped('order_line')) == 0)
+        orderless_partners = to_remove.mapped('partner_id')
+        to_remove.unlink()
+        # Remove partners with no orders in the current document and no
+        # historic orders.
+        # TODO check this query is efficient.
+        domain = [('id', 'in', orderless_partners.mapped('id')),
+                  ('sale_order_ids', '=', False)]
+        partners_with_no_historic_orders = Partner.search(domain)
+        partners_with_no_historic_orders.unlink()
+
+        # TODO
+        # It would be good to do the note creation here.  Would need to match
+        # requests vs actually created orders and lines to report them.
+
         # Automatically confirm sale orders, if applicable
         if self._auto_confirm:
-            SaleRequestRecord = self.sale_request_record_model(doc)
-            reqs = SaleRequestRecord.search([('doc_id', '=', doc.id)])
-            for r, sales in reqs.mapped('sale_id').batched(self.BATCH_CONFIRM):
+            for r, sales in reqs.exists().mapped('sale_id').batched(self.BATCH_CONFIRM):
```

Comment:
Is the `exists()` needed ? we are not deleting the sale order requests. 

Could we use `sales - to_remove` instead of `reqs.exists().mapped('sale_id')`?

Reply:
Done, please review.